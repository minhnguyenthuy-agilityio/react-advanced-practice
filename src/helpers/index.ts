export * from "./testing";
export * from "./validate";
