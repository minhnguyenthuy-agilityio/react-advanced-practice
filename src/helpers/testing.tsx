import { ReactNode } from "react";

import { QueryClient, QueryClientProvider } from "react-query";

import { render } from "@testing-library/react";

import { CHAKRA_THEME } from "@/themes";

import { BrowserRouter as Router } from "react-router-dom";

import { ChakraProvider } from "@chakra-ui/react";

export const renderWithProvider = (children: ReactNode) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        staleTime: 60 * 1000,
      },
    },
  });

  return render(
    <QueryClientProvider client={queryClient}>
      <Router>
        <ChakraProvider theme={CHAKRA_THEME}>{children}</ChakraProvider>
      </Router>
    </QueryClientProvider>,
  );
};

export const wrapper = ({ children }: { children: ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

export * from "@testing-library/react";
export { renderWithProvider as render };
