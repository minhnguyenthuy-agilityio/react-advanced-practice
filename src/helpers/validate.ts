import { REGEXP } from "@/constants";

export const isValidImageUrl = (value: string | undefined) => {
  return !value || REGEXP.IMAGE.test(value);
};
