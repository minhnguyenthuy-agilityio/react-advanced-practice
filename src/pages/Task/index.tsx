import { Suspense } from "react";

import { Box, Flex, Heading, Spinner } from "@chakra-ui/react";

import { useParams } from "react-router-dom";

import { FilterIcon } from "@/assets/icons";

import { PROJECTS } from "@/mocks";

import { OPTIONSFILTER, Status } from "@/constants";

import { TaskList, MenuFilter } from "@/components";

import { useTaskList, useTaskListStore } from "@/hooks";

export const TaskPage = () => {
  const { projectId } = useParams();
  const { keyword } = useTaskListStore();
  const { data: tasks, isLoading } = useTaskList(
    projectId ? projectId : "pro00",
  );

  // Currently there is no homepage so the default page will be the Mobile App project with the url /
  // List project is now mock data
  const projectName = projectId
    ? PROJECTS.find((project) => project.projectId === projectId)?.name
    : "Mobile App";

  const taskListFilter = () => {
    const statusArray = [Status.Todo, Status.Onprogress, Status.Done];

    return statusArray.map((status) => {
      let tasksByStatus = tasks?.filter((task) => task.status === status) || [];

      if (keyword) {
        tasksByStatus = tasksByStatus.filter((task) =>
          task.name.toLowerCase().includes(keyword.toLowerCase()),
        );
      }

      return <TaskList key={status} status={status} tasks={tasksByStatus} />;
    });
  };

  return (
    <Suspense
      fallback={
        <Spinner
          thickness="4px"
          speed="0.65s"
          emptyColor="gray.200"
          color="primary.300"
          size="md"
        />
      }
    >
      <Box px="40px" pt="35px" w="full" h="full" overflowY="hidden">
        <Flex justifyContent="space-between" alignItems="center">
          <Heading size="xl">{projectName}</Heading>
          <Flex gap={2}>
            <MenuFilter
              buttonName="Filter"
              options={OPTIONSFILTER}
              icon={FilterIcon}
            />
            <MenuFilter buttonName="Today" options={OPTIONSFILTER} />
          </Flex>
        </Flex>
        <Flex justifyContent="space-between" mt="30px" h="full">
          {isLoading ? (
            <Spinner
              thickness="4px"
              speed="0.65s"
              emptyColor="gray.200"
              color="primary.300"
              size="md"
              m="auto"
            />
          ) : (
            taskListFilter()
          )}
        </Flex>
      </Box>
    </Suspense>
  );
};
