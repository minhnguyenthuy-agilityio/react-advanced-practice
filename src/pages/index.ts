export * from "./Main";
export * from "./Task";
