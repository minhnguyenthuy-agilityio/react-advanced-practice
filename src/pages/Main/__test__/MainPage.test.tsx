import { render } from "@/helpers";
import { Page } from "@/pages";

describe("Main Page", () => {
  it("matchs snapshoot", () => {
    const element = render(<Page />);
    expect(element).toMatchSnapshot();
  });
});
