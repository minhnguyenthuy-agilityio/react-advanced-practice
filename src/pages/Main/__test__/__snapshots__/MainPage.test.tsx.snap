// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page matchs snapshoot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <article>
        <section
          class="storybook-page"
        >
          <h2>
            Pages in Storybook
          </h2>
          <p>
            We recommend building UIs with a
             
            <a
              href="https://componentdriven.org"
              rel="noopener noreferrer"
              target="_blank"
            >
              <strong>
                component-driven
              </strong>
            </a>
             
            process starting with atomic components and ending with pages.
          </p>
          <p>
            Render pages with mock data. This makes it easy to build and review page states without needing to navigate to them in your app. Here are some handy patterns for managing page data in Storybook:
          </p>
          <ul>
            <li>
              Use a higher-level connected component. Storybook helps you compose such data from the "args" of child component stories
            </li>
            <li>
              Assemble data in the page component from your services. You can mock these services out using Storybook.
            </li>
          </ul>
          <p>
            Get a guided tutorial on component-driven development at
             
            <a
              href="https://storybook.js.org/tutorials/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Storybook tutorials
            </a>
            . Read more in the
             
            <a
              href="https://storybook.js.org/docs"
              rel="noopener noreferrer"
              target="_blank"
            >
              docs
            </a>
            .
          </p>
          <div
            class="tip-wrapper"
          >
            <span
              class="tip"
            >
              Tip
            </span>
             Adjust the width of the canvas with the
             
            <svg
              height="10"
              viewBox="0 0 12 12"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
              >
                <path
                  d="M1.5 5.2h4.8c.3 0 .5.2.5.4v5.1c-.1.2-.3.3-.4.3H1.4a.5.5 0 01-.5-.4V5.7c0-.3.2-.5.5-.5zm0-2.1h6.9c.3 0 .5.2.5.4v7a.5.5 0 01-1 0V4H1.5a.5.5 0 010-1zm0-2.1h9c.3 0 .5.2.5.4v9.1a.5.5 0 01-1 0V2H1.5a.5.5 0 010-1zm4.3 5.2H2V10h3.8V6.2z"
                  fill="#999"
                  id="a"
                />
              </g>
            </svg>
            Viewports addon in the toolbar
          </div>
        </section>
      </article>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <article>
      <section
        class="storybook-page"
      >
        <h2>
          Pages in Storybook
        </h2>
        <p>
          We recommend building UIs with a
           
          <a
            href="https://componentdriven.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            <strong>
              component-driven
            </strong>
          </a>
           
          process starting with atomic components and ending with pages.
        </p>
        <p>
          Render pages with mock data. This makes it easy to build and review page states without needing to navigate to them in your app. Here are some handy patterns for managing page data in Storybook:
        </p>
        <ul>
          <li>
            Use a higher-level connected component. Storybook helps you compose such data from the "args" of child component stories
          </li>
          <li>
            Assemble data in the page component from your services. You can mock these services out using Storybook.
          </li>
        </ul>
        <p>
          Get a guided tutorial on component-driven development at
           
          <a
            href="https://storybook.js.org/tutorials/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Storybook tutorials
          </a>
          . Read more in the
           
          <a
            href="https://storybook.js.org/docs"
            rel="noopener noreferrer"
            target="_blank"
          >
            docs
          </a>
          .
        </p>
        <div
          class="tip-wrapper"
        >
          <span
            class="tip"
          >
            Tip
          </span>
           Adjust the width of the canvas with the
           
          <svg
            height="10"
            viewBox="0 0 12 12"
            width="10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M1.5 5.2h4.8c.3 0 .5.2.5.4v5.1c-.1.2-.3.3-.4.3H1.4a.5.5 0 01-.5-.4V5.7c0-.3.2-.5.5-.5zm0-2.1h6.9c.3 0 .5.2.5.4v7a.5.5 0 01-1 0V4H1.5a.5.5 0 010-1zm0-2.1h9c.3 0 .5.2.5.4v9.1a.5.5 0 01-1 0V2H1.5a.5.5 0 010-1zm4.3 5.2H2V10h3.8V6.2z"
                fill="#999"
                id="a"
              />
            </g>
          </svg>
          Viewports addon in the toolbar
        </div>
      </section>
    </article>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
